src/main/java/org/jabref/logic/shared/DBMSSynchronizer.java

```diff
@@ -221,7 +231,8 @@ public void synchronizeLocalDatabase() {
 
         if (!entriesToInsertIntoLocalDatabase.isEmpty()) {
             // in case entries should be added into the local database, insert them
-            bibDatabase.insertEntries(dbmsProcessor.getSharedEntries(entriesToInsertIntoLocalDatabase), EntriesEventSource.SHARED);
+            taskExecutor.runInFXThread(() ->
```

Comment:
Is this really necessary? I would say it's the job of the ui to make sure it handles notifications about new entries on the right thread, and not of the `logic`. In fact, the logic shouldn't even know about the concept of fx-threads.

Reply:
Yes, while I agree with you that from an architectural standpoint this looks odd, the key problem is that if you insert anything in the bib database or modify a field, the GUI observables in the entry editor or maintable are called as well because they are bound to the bib database object.

Reply:
For this situation, we have the "forUi" helper
https://github.com/JabRef/jabref/blob/de114d520ac0a8832bb2eacae29f1745c2f0113a/src/main/java/org/jabref/gui/util/BindingsHelper.java#L165-L170

Reply:
Yes, but I have encountered this before, this leads to a deadlock under certain conditions:
Comment that line out and connect to the shared SQL library in #8485 
See here. I can only solve this by adding a timeout in the UI Thread list

![grafik](https://user-images.githubusercontent.com/320228/153651836-e50534df-9b9a-42fa-aec1-bc8b9549d26f.png)

Reply:
> NOTE: applications should avoid flooding JavaFX with too many pending Runnables. Otherwise, the application may become unresponsive. Applications are encouraged to batch up multiple operations into fewer runLater calls. Additionally, long-running operations should be done on a background thread where possible, freeing up the JavaFX Application Thread for GUI operations. 

That might be the reason. How many source changes does the database trigger?

Reply:
**This is not the case here** it's not a problem that the UI is getting slow, we have a kind of **deadlock** blocking situation here because the underlying list a SynchronizedLIst which holds a mutex for every operation, even iteration.


1. BackgroundTask here is inserting entries into the bibdatabase, into the local collection 
`    private final ObservableList<BibEntry> entries = FXCollections.synchronizedObservableList(FXCollections.observableArrayList(BibEntry::getObservables));`


2. Now we are in UIThreadList and want to post the changes on the UI Thread, so far so good 
3. We acquire the FX thread and go forward and fire our value change. 
4. We call firevalueChange and end up in the MapBackedList and try to get the changes from the original list (the SycnchronizedLIst)
5. We call get and see that the mutex is hold by the Background thread already. So we wait...

FX Thread:
```

Thread [JavaFX Application Thread] (Suspended)	
	waiting for: FXCollections$SynchronizedObservableList<T>  (id=139)	
		owned by: Thread [pool-2-thread-4] (Stepping)	
	FXCollections$SynchronizedObservableList<T>(FXCollections$SynchronizedList<T>).get(int) line: 1124	
	FXCollections$UnmodifiableObservableListImpl<T>.get(int) line: 970	
	MappedBackedList<E,F>.sourceChanged(Change<? extends F>) line: 63	
	MappedBackedList<E,F>(TransformationList<E,F>).lambda$getListener$0(ListChangeListener$Change) line: 106	
	0x00000007c03b6638.onChanged(ListChangeListener$Change) line: not available	
	WeakListChangeListener<E>.onChanged(Change<? extends E>) line: 88	
	ListListenerHelper$SingleChange<E>.fireValueChangedEvent(Change<? extends E>) line: 164	
	ListListenerHelper<E>.fireValueChangedEvent(ListListenerHelper<E>, Change<? extends E>) line: 73	
	UiThreadList<T>(ObservableListBase<E>).fireChange(Change<? extends E>) line: 239	
	UiThreadList<T>.lambda$0(ListChangeListener$Change, CountDownLatch) line: 26	
	0x00000007c0858a08.run() line: not available [local variables unavailable]	
	PlatformImpl.lambda$runLater$10(Runnable) line: 457	
	0x00000007c018a830.run() line: not available	
	AccessController.executePrivileged(PrivilegedAction<T>, AccessControlContext, Class<?>) line: 776	
	AccessController.doPrivileged(PrivilegedAction<T>, AccessControlContext) line: 399	
	PlatformImpl.lambda$runLater$11(Runnable, AccessControlContext) line: 456	
	0x00000007c01897d0.run() line: not available	
	InvokeLaterDispatcher$Future.run() line: 96	

```


Backtround Thread stack:

```
Thread [pool-2-thread-4] (Suspended (breakpoint at line 24 in UiThreadList))	
	owns: FXCollections$SynchronizedObservableList<T>  (id=139)	
	owns: BibDatabase  (id=114)	
	UiThreadList<T>.sourceChanged(Change<? extends T>) line: 24	
	UiThreadList<T>(TransformationList<E,F>).lambda$getListener$0(ListChangeListener$Change) line: 106	
	0x00000007c03b6638.onChanged(ListChangeListener$Change) line: not available	
	WeakListChangeListener<E>.onChanged(Change<? extends E>) line: 88	
	ListListenerHelper$SingleChange<E>.fireValueChangedEvent(Change<? extends E>) line: 164	
	ListListenerHelper<E>.fireValueChangedEvent(ListListenerHelper<E>, Change<? extends E>) line: 73	
	FXCollections$UnmodifiableObservableListImpl<T>(ObservableListBase<E>).fireChange(Change<? extends E>) line: 239	
	FXCollections$UnmodifiableObservableListImpl<T>.lambda$new$0(ListChangeListener$Change) line: 963	
	0x00000007c0187290.onChanged(ListChangeListener$Change) line: not available	
	WeakListChangeListener<E>.onChanged(Change<? extends E>) line: 88	
	ListListenerHelper$Generic<E>.fireValueChangedEvent(Change<? extends E>) line: 329	
	ListListenerHelper<E>.fireValueChangedEvent(ListListenerHelper<E>, Change<? extends E>) line: 73	
	FXCollections$SynchronizedObservableList<T>.lambda$new$0(ListChangeListener$Change) line: 1217	
	0x00000007c0657db0.onChanged(ListChangeListener$Change) line: not available	
	WeakListChangeListener<E>.onChanged(Change<? extends E>) line: 88	
	ListListenerHelper$SingleChange<E>.fireValueChangedEvent(Change<? extends E>) line: 164	
	ListListenerHelper<E>.fireValueChangedEvent(ListListenerHelper<E>, Change<? extends E>) line: 73	
	ObservableListWrapper<E>(ObservableListBase<E>).fireChange(Change<? extends E>) line: 239	
	ListChangeBuilder<E>.commit() line: 482 [local variables unavailable]	
	ListChangeBuilder<E>.endChange() line: 541 [local variables unavailable]	
	ObservableListWrapper<E>(ObservableListBase<E>).endChange() line: 211	
	ObservableListWrapper<E>(ModifiableObservableListBase<E>).addAll(Collection<? extends E>) line: 109	
	FXCollections$SynchronizedObservableList<T>(FXCollections$SynchronizedList<T>).addAll(Collection<? extends T>) line: 1088	
	BibDatabase.insertEntries(List<BibEntry>, EntriesEventSource) line: 216	
	DBMSSynchronizer.synchronizeLocalDatabase() line: 243	
	DBMSSynchronizer.lambda$4() line: 187	
	0x00000007c0852528.run() line: not available	
	BackgroundTask$2.call() line: 68	
	BackgroundTask$2.call() line: 1	
	DefaultTaskExecutor$1.call() line: 162	
	Task$TaskCallable<V>.call() line: 1426	
	DefaultTaskExecutor$1(FutureTask<V>).run() line: 264	
	Executors$RunnableAdapter<T>.call() line: 539	
	FutureTask<V>.run() line: 264	
	ThreadPoolExecutor.runWorker(ThreadPoolExecutor$Worker) line: 1136	
	ThreadPoolExecutor$Worker.run() line: 635	
	Thread.run() line: 833	
```

Reply:
Nice analysis, thanks!

What I don't understand is why does the background thread pauses in the UiThreadList? Shouldn't it go on, sync the rest of the changes and finally give the synced entry list free?

Maybe it works if one adds the `forUi` call at the last position here https://github.com/JabRef/jabref/blob/ba5beb2f330b984d9653db7fdeae81bbf2e8ac28/src/main/java/org/jabref/gui/maintable/MainTableDataModel.java#L57 (or for the entriesFiltered, if otherwise the search no longer works). Then all the syncs / updates of the mapped database would still run in the background thread.

Reply:
the BackgroundThread only indirectly pauses in the UIList because 
1. When calling add() on the SynchronizedLIst, Background Thread acquires a mutex for the whole list.
2. it's waiting for the CountDownLatch that never happens; this is why I tested to solve it with a timeout in the latch.

```
     CountDownLatch latch = new CountDownLatch(1);
            Platform.runLater(() -> {
                fireChange(change);
                latch.countDown(); //never gets called, because we are stil in fireChange  and bocked
            });

            try {
                latch.await();
               // boolean latched = latch.await(3L, TimeUnit.SECONDS);
               // LOGGER.debug("Result from latch {}", latched); // if this is false, it was canceled by the timeout

            } catch (InterruptedException e) {
                LOGGER.error("Error while running on JavaFX thread", e);
            } finally {
                latch.countDown();
            }
```

Reply:
Mhh, maybe it's a good idea to step back and try to come up with a general solution to these problems?
Off the top of my head, we have two ways to handle background threads that result in updates to the UI:
- Background threads are only allowed to calculate things, but never change any of the data model classes directly.
- Background threads are allowed to change data models, but at every point where the data model classes are bind to the UI we make sure that the updates are happening on the JavaFX thread (that essentially only concerns the main table, group view and entry editor).

I think both ways have advantages and disadvantages. I don't have a strong preference right now, do you? But I do think it would make sense to decide which approach we take and then do changes accordingly.

Reply:
From an architectural standpoint, updating the model should not have direct influences of the UI. The UI should take care that it is updating things in the right thread. This is more or less our current architecture.  (2nd approach).

The key problem is more that we have **SynchronizedList** which is bound to the UI.
It has a mutex on all methods which will be held when elements are inserted. So while Elements are inserted a lock is held. Firing Changes on the FX Thread will always access the underlying list. 
I will see if this work with the forUI for groups as well